File: stringMatch/LongestCharacters3/LongestCharacters3.cc
/**
 * 3. Longest Substring Without Repeating Characters
 * Given a string s, find the length of the longest substring without repeating characters.
 * 
 * Example
 * Input: s = "abcabcbb"
 * Output: 3
 * Explanation: The answer is "abc", with the length of 3.
 * */
#include<string>
#include<algorithm>
#include<unordered_map>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> windows;
        int left = 0, right = 0;
        int res = 0;
        while (right < s.size()){
            char c = s[right];
            right ++;
            windows[c]++;
            //窗口元素大于1 说明有重复 缩小窗口
            while (windows[c] > 1){
                char d = s[left];
                left ++;
                windows[d]--;
            }
            res = max(res, right - left);
        }
        return res;
    }
};

File: stringMatch/LongestCharacters3/LongestCharacters3.go
/*
3. 无重复字符的最长子串
给定一个字符串，请你找出其中不含有重复字符的 最长子串 的长度。
输入: s = "abcabcbb"
输出: 3
解释: 因为无重复字符的最长子串是 "abc"，所以其长度为 3。
*/

package LongestCharacters3

func lengthOfLongestSubstring(s string) int {
	windows := make(map[byte]int, len(s))
	var (
		left, right int
		maxLen      int
	)
	for right < len(s) {
		c := s[right]
		right++
		windows[c]++
		for windows[c] > 1 {
			d := s[left]
			left++
			windows[d]--
		}
		if right-left > maxLen {
			maxLen = right - left
		}
	}
	return maxLen
}
